### DISCUSSION TITLE
binary dilation, thresholding

### DISCUSSION THREAD

**Discussion by mohsen-saki** [libvips/libvips#2998] [Q&A] [answered]
Hello @jcupitt /Community,

Thank you for taking your time to patiently respond to the questions.

I have been trying to implement an image process pipeline in pyvips borrowed from skimage.

The purpose is to filter through medical image tiles and remove those with no or less tissue (thresholding)

in skimage it is:

```
tile = rgb2gray(rgba2rgb(tile))
tile = canny(tile)
tile = binary_dilation(tile, disk(6))
pct = tile.mean()
```
then compare `pct` with any margin and keep or leave the tile.

In pyvips I did as below:

```
thresh_mean = tile.colourspace('b-w').avg() 
mask_array = [[128.0, 255.0, 128.0], [255.0, 255.0, 255.0], [128.0, 255.0, 128.0]] 
mask = pyvips.Image.new_from_array(mask_array) 
tile = (tile.colourspace('b-w').canny().dilate(mask) > thresh_mean).ifthenelse(0,1) 
pct = tile.avg() 
```

Which seems is doing the same thing, but the performance does not seem robust enough. For example, when a tile contains only the background, pyvips code cannot properly handle the situation.

original tile:
![download](https://user-images.githubusercontent.com/47094390/185730311-c179aecf-4c1b-4754-9169-fa8d50d9477c.png)

skimage output: pct = ~ 0.05
![download](https://user-images.githubusercontent.com/47094390/185730364-fa5bb89b-81f2-44aa-89a7-15b3f0f90097.png)

pyvips output: pct = ~ 0.46
![download](https://user-images.githubusercontent.com/47094390/185730805-6c03862d-535f-4113-9589-4e7bfceddcce.png)

For the tiles in which the foreground is apparent, it does almost the same but still a bit dodgy specially when the colour tone of foreground and background is relatively close.

Just in case, using simpler still effective implementations like below have not been helping much because either they do not perform well with background tiles or a single threshold like `200` will not handle all the whole slides due to huge variation in color tones.
```
tile = (tile.colourspace('b-w') > thresh_mean).ifthenelse(0,1)
tile = (tile.colourspace('b-w') > 200).ifthenelse(0,1)
```

So, basically my question is what is a roubust way to handle thresholding in pyvips/linvips?

I would like to break it down to:

Is it possible to perform binary masking? I know I can generate circular binary mask using `mask_ideal` but not usable with `dilate()` I suppose!

Is there an easier way to create larger `mask_arrays` (like above in the code). I copy-pasted that from libvips docs. Have not seen a function to create one!

Much Appreciated,
Mohsen

**Comment by jcupitt** [comment]
Hi @mohsen-saki,

Could you upload a couple of tricky sample background tiles? You might need to adjust the canny parameters a bit.

When I've done "discard bg tiles" before I've found simple histogram tests like "discard tiles with x% of pixels withing y% of the background" to work well, but I suppose that's not working with your images. I imagine it must depend quite a bit on the scanner and the slides.

I've also used things like `image - image.gaussblur(10)` as a preprocessing step to remove very low frequency changes in the image caused by lighting differences. That might be enough to make a simple threshold work. 

You could write this:

```python
tile = (tile.colourspace('b-w').canny().dilate(mask) > thresh_mean).ifthenelse(0,1) 
pct = tile.avg() 
```

As:

```python
tile = tile.colourspace('b-w').canny().dilate(mask) < thresh_mean
pct = tile.avg() / 255
```

Since bool operations make 255/0 pixels for TRUE and FALSE.

You can also use a rank filter instead of dilate -- it works well as a grey-level morphology. Maybe:

```python
tile = tile.colourspace('b-w').canny().rank(3, 3, 8) < thresh_mean
pct = tile.avg() / 255
```

It is a bit slower though.

**Comment by jcupitt** [comment]
You can pass a mask from `mask_ideal` to `dilate`, you just have to scale the range. For example:

```
$ python3
Python 3.10.4 (main, Jun 29 2022, 12:14:53) [GCC 11.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyvips
>>> x = pyvips.Image.mask_ideal(11, 11, 1, optical=True, reject=True)
>>> x.numpy()
array([[0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0.],
       [0., 0., 1., 1., 1., 1., 1., 1., 1., 0., 0.],
       [0., 1., 1., 1., 1., 1., 1., 1., 1., 1., 0.],
       [0., 1., 1., 1., 1., 1., 1., 1., 1., 1., 0.],
       [0., 1., 1., 1., 1., 1., 1., 1., 1., 1., 0.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [0., 1., 1., 1., 1., 1., 1., 1., 1., 1., 0.],
       [0., 1., 1., 1., 1., 1., 1., 1., 1., 1., 0.],
       [0., 1., 1., 1., 1., 1., 1., 1., 1., 1., 0.],
       [0., 0., 1., 1., 1., 1., 1., 1., 1., 0., 0.],
       [0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0.]], dtype=float32)
>>> y = (x * 128 + 128).cast("uchar")
>>> y.numpy()
array([[128, 128, 128, 128, 128, 255, 128, 128, 128, 128, 128],
       [128, 128, 255, 255, 255, 255, 255, 255, 255, 128, 128],
       [128, 255, 255, 255, 255, 255, 255, 255, 255, 255, 128],
       [128, 255, 255, 255, 255, 255, 255, 255, 255, 255, 128],
       [128, 255, 255, 255, 255, 255, 255, 255, 255, 255, 128],
       [255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255],
       [128, 255, 255, 255, 255, 255, 255, 255, 255, 255, 128],
       [128, 255, 255, 255, 255, 255, 255, 255, 255, 255, 128],
       [128, 255, 255, 255, 255, 255, 255, 255, 255, 255, 128],
       [128, 128, 255, 255, 255, 255, 255, 255, 255, 128, 128],
       [128, 128, 128, 128, 128, 255, 128, 128, 128, 128, 128]],
      dtype=uint8)
>>> 

```

**Reply by jcupitt** [reply]
pyvips will display image values for the matrix type, but the numpy repr is prettier. For example:

```
>>> x
<pyvips.Image 11x11 float, 1 bands, fourier>
>>> y = x.copy(interpretation="matrix")
>>> y
[[0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0], [0.0, 0.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 0.0, 0.0], [0.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 0.0], [0.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 0.0], [0.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 0.0], [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0], [0.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 0.0], [0.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 0.0], [0.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 0.0], [0.0, 0.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 0.0, 0.0], [0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0]]
```

Perhaps we should improve this.

**Comment by mohsen-saki** [comment, marked as answer]
> Could you upload a couple of tricky sample background tiles? You might need to adjust the canny parameters a bit.

[Here (on Gdrive)](https://drive.google.com/file/d/1T6kIF1dvQ1Ka3RgOq_ZoorhMBYQuKiKZ/view?usp=sharing) are a couple of 10k or 15k cropped slides which are to some extent troublesome. specifically `slide_01` is a good one showing my implementation of `pyvips` will not produce as expected (p.s.: I use 2kx2k tiles for my trials).

Thanks to the tips, I do as follows:
```
mask = pyvips.Image.mask_ideal(13, 13, 1, optical=True, reject=True)
mask = (mask * 128 + 128).cast("uchar")
tile = tile.colourspace('b-w').canny().dilate(mask)
pct = tile.avg()/255
```
> When I've done "discard bg tiles" before I've found simple histogram tests like "discard tiles with x% of pixels withing y% of the background" to work well.

I would like to have a look at your implementation of histogram based thresholding if not too much trouble to share a link. The problem with morphology operations like `canny()` and `dilate()`- even if perfectly accurate- is that they are still relatively slow and might not be very practical.

> I've also used things like image - image.gaussblur(10) as a preprocessing step to remove very low frequency changes 

I tried a couple of cases and the result was not consistent. Will keep playing around with different `sigma` and report back if anything interesting.

I am currently trying to utilise the saturation layer of a HSV colour model. It is surprisingly good and noticeably quick, but does not generalise well like others as there are some cases of deviation.

Implemented as below:

```
h,s,v = tile.sRGB2HSV().bandsplit()
tile = s > s.avg()
pct = tile.avg()/255
```

**Reply by mohsen-saki** [reply]
Just in case anyone ended up here for the same reason, adding parameter `precision='integer'` to the `canny()` operation will make the pipeline almost identical to `skimage` version. 

one more thing- as a general rule for most medical slides- green channel happens to contain most of  structural data and probably working better than `grayscale` here. Having said that, below lines of code has been working good so far:

```
mask = pyvips.Image.mask_ideal(13, 13, 1, optical=True, reject=True)
mask = (mask * 128 + 128).cast("uchar")
r,g,b = tile.bandsplit()
tile = g.canny(precision='integer').dilate(mask)
return tile.avg()/255
```

**The bonus is, it is almost 10x faster than `skimage` pipeline.**

I still believe there should be even faster implementation because for the task of "discard tiles with more background", accuracy is not the key factor and perhaps there could be a trade off in favour of faster operation.

Hope @jcupitt could find some time and comment `histogram` based implementation.

**Reply by jcupitt** [reply]
That's great!

You can use:

```
g = tile[1]
```

To take the green channel. It'll be slightly quicker. Dropping the mask radius would speed things up too.

**Reply by mohsen-saki** [reply]
> Dropping the mask radius would speed things up too.

I have already played around with that and lower than `11` it will run into the same problem! Will produce tiles which are more of background rather than good stuff.

**Comment by mohsen-saki** [comment]
@jcupitt  , One more issue that you may find it interesting to look into is that some of `pyvips` operations will break a multi-process thread.

I used to run the pipeline through a multi-process and get some boost re time, but `dilate()` will break a python `ProcessPoolExecutor` while `ThreadPoolExecutor` is OK.

I had the same issue with `sobel()` before.

The error is:

`BrokenProcessPool: A process in the process pool was terminated abruptly while the future was running or pending.`

**Reply by jcupitt** [reply]
You need the `import pyvips` to run in the main thread, could that be it? 

If you could make a quick test case I could have a look.

**Reply by mohsen-saki** [reply]
> You need the `import pyvips` to run in the main thread, could that be it?

Should not be that one.

Here I created a notebook [pyvips_test_case](https://www.kaggle.com/code/sakima/pyvips-test-case) which should regenerate the issue.

Need to mention that I am using `fastai` library on top of `PoolExecutor`, but I suspect `pyvips` is interrupting the process as I have been running this code many times with other simpler pipelines of `pyvips` with no problem.

here is `fastai` [docs](https://fastcore.fast.ai/parallel.html#threadpoolexecutor) for `parallel()` and here is the [source code](https://github.com/fastai/fastcore/blob/master/fastcore/parallel.py).

Thanks

**Reply by mohsen-saki** [reply]
Sorry! better put a pause on this! The behaviour seems a bit random! I am investigating!

**Reply by jcupitt** [reply]
You've got this at the top level:

```python 
mask = pyvips.Image.mask_ideal(13, 13, 1, optical=True, reject=True)
mask = (mask * 128 + 128).cast("uchar")
```

But that's being read (and recomputed) by each thread. Try:

```python 
mask = pyvips.Image.mask_ideal(13, 13, 1, optical=True, reject=True)
mask = (mask * 128 + 128).cast("uchar").copy_memory()
```

Now it'll run that pipeline once during startup and make an image in memory, then use that single pointer (no evaluation) in all of the worker threads.

**Reply by jcupitt** [reply]
I expect you know, but you could write this:

```python
def get_tiles(slide, t_size=2000):
    slide = pyvips.Image.new_from_file(slide)
    n_across =  slide.width // t_size
    n_down = slide.height // t_size

    for h in range(0, n_down):
        for w in range (0, n_across):
            tile = slide.crop(w*t_size, h*t_size, t_size, t_size)
            yield tile, (t_size, (w,h))
```

As:

```python
def get_tiles(slide, t_size=2000):
    slide = pyvips.Image.new_from_file(slide)
    return (slide.crop(x, y, min(t_size, slide.width - x), min(t_size, slide.height - y)), (t_size, (x, y))
            for y in range(0, slide.height, t_size)
            for x in range(0, slide.width, t_size))
```

You could pad edge tiles to 2k x 2k as well, I don't know if that's an issue.

**Reply by jcupitt** [reply]
You could also write:

```
tile = tile.resize(0.25, kernel='VIPS_KERNEL_LINEAR')
```

As:

```
tile = tile.resize(0.25, kernel='linear')
```

**Reply by mohsen-saki** [reply]
Tanks again @jcupitt for the the practical tips.

**I have to say it! I believe `pyvips (/libvips)` deserve a better and more elaborate documentation!** Hey! to the community off course! :)

> You could pad edge tiles to 2k x 2k as well, I don't know if that's an issue.

To be honest!  I've been so focused on speeding up while filtering tiles through that I have not thought about that!

I suppose you mean for edge tiles which are missed in this tiling process? and probably using `gravity()`?

Have not used yet! Could it be piped to crop?

**Reply by mohsen-saki** [reply]
Whoops! I got it!
With below implementation, it would produce edge tiles anyway and definitely requires `zeros` padding! 

```
def get_tiles(slide, t_size=2000):
    slide = pyvips.Image.new_from_file(slide)
    return (slide.crop(x, y, min(t_size, slide.width - x), min(t_size, slide.height - y)), (t_size, (x, y))
            for y in range(0, slide.height, t_size)
            for x in range(0, slide.width, t_size))
```

I will do:

`slide.crop(...).gravity('centre', t_size, t_size)`

Right?

**Reply by jcupitt** [reply]
That's it. Maybe 'north-west' would be better than `centre`? Gravity will pad with 0 by default, but it'll do white (and tile) as well.

**Reply by jcupitt** [reply]
PRs with doc improvements are always very welcome, of course :)
